DHCP reservation actions - list/add/delete

kgaikwad opened the pull request.

kgaikwad commented on definitions/features/foreman_proxy.rb:
```diff
@@ -3,19 +3,62 @@ class Features::ForemanProxy < ForemanMaintain::Feature
   end
 
+  def list_reservations(subnet)
+    resp = dhcp_restclient["/#{subnet}"].get { |response, _request, _result| response }
```
@iNecas ,

I need your suggestion for these API requests.
Whether it would be good - instead of placing API calls in `foreman_proxy` feature If I extract out these API calls into `utils/api` folder to add class files in which it will do API interface.

for example, in this case I will add dhcp_reservation.rb under utils/api folder to add methods fetch_list/add/delete.


```
module ForemanMaintain
  module Utils
    module Api
      class DhcpReservation
        include ForemanMaintain::Concerns::SystemHelpers

        attr_reader :dhcp_api_resource

        def self.instance(base_url, options = {})
          @instance ||= new(base_url, options)
        end

        def initialize(base_url, options = {})
          @dhcp_api_resource = RestClient::Resource.new(base_url, options)
        end

        def fetch_list(subnet)
          resp = dhcp_api_resource["/#{subnet}"].get { |response, _request, _result| response }
          handle_api_response(resp, true)
        end
     end
    end
end
```

iNecas replied:
I can imagine having ForemanMaintain::Utils::SmartProxyApi and using that from foreman_proxy feature

swapab replied:
Another option I think could be `ForemanMaintain::Utils::DhcpReservation` with instance method `#list #add #delete`

iNecas commented on definitions/features/foreman_proxy.rb:
```diff
@@ -3,19 +3,63 @@ class Features::ForemanProxy < ForemanMaintain::Feature
   def initialize
     @dhcpd_conf_file = '/etc/dhcp/dhcpd.conf'
     @cert_path = ForemanMaintain.config.foreman_proxy_cert_path
+    @request_url = "https://#{hostname}:9090/dhcp"
```
The port is also configurable: we should not hardcode it here

swapab replied:
Would it be good to use URI library[1] for URL(s) instead of plain strings?
[1]https://ruby-doc.org/stdlib-2.1.2/libdoc/uri/rdoc/URI/HTTP.html

iNecas commented:
I guess we should also make this available via `foreman-maintain dhcp list/add/delete`, so that the user doesn't have to go via `advanced procedure run`

kgaikwad commented:
@iNecas ,
I will commit dhcp command code which was added by me for testing purpose on my local machine.

swapab commented on Gemfile:
```diff
@@ -3,6 +3,7 @@ source 'https://rubygems.org'
 gemspec
 
 gem 'pry'
+gem 'rest-client'
```
Should this be a `development_dependency`?

swapab commented on definitions/features/foreman_proxy.rb:
```diff
@@ -3,19 +3,63 @@ class Features::ForemanProxy < ForemanMaintain::Feature
+
   private
 
+  def load_default_ssl_cert_options
```
I would prefer an SSL utility class that accepts following file paths as default arguments.
And `SSL#to_h` would return the desired hash returned here.
Sounds better?

kgaikwad commented:
@iNecas, @swapab  
Done with modifications as per suggestions.
- Created two Utils Classes
  1. DhcpReservation
  2. SmartProxySSL
- Made `foreman-proxy-fqdn` and `foreman-proxy-port` configurable 
Default values set as below =>
`foreman-proxy-fqdn=#{hostname}` 
`foreman-proxy-port=9090`
- Created DHCP as a separate command using [abstract_procedure_command.rb](https://github.com/theforeman/foreman_maintain/blob/master/lib/foreman_maintain/cli/advanced/procedure/abstract_procedure_command.rb)
- Added `rest-client` in `.gemspec` file
- Applied `for_feature`, `tags` and `label` to dhcp list/add/delete procedures
- Added `confine` block to check cert files present to execute check/procedure for feature `foreman-proxy`

Ready for review.

kgaikwad commented:
@iNecas ,
Modified sub-command names for main `dhcp` command. 
Now it will show,

**1. DHCP Main command:**

```
[root@satellite1 foreman_maintain]# ./bin/foreman-maintain dhcp --help
Usage:
    foreman-maintain dhcp [OPTIONS] SUBCOMMAND [ARG] ...

Parameters:
    SUBCOMMAND                    subcommand
    [ARG] ...                     subcommand arguments

Subcommands:
    reservation-add               add DHCP reservation to subnet
    reservation-delete            delete DHCP reservation using subnet
    reservation-list              list DHCP reservations for subnet

Options:
    -h, --help                    print help
```

**2.  Run dhcp-reservation procedures using `advanced procedure run`:**

```
[root@satellite1 foreman_maintain]# ./bin/foreman-maintain advanced procedure run --help
Usage:
    foreman-maintain advanced procedure run [OPTIONS] SUBCOMMAND [ARG] ...

Parameters:
    SUBCOMMAND                    subcommand
    [ARG] ...                     subcommand arguments

Subcommands:
    dhcp-reservation-add          add DHCP reservation to subnet
    dhcp-reservation-delete       delete DHCP reservation using subnet
    dhcp-reservation-list         list DHCP reservations for subnet
```

swapab commented on definitions/features/foreman_proxy.rb:
```diff
@@ -1,21 +1,58 @@
+
   def dhcp_curl_cmd
     "curl -w '\n%{http_code}' -slient -ks --cert #{cert_path}/client_cert.pem \
       --key #{cert_path}/client_key.pem \
```
Port `9090` is configurable I suppose?

swapab replied:
https://github.com/theforeman/foreman_maintain/pull/84/files#diff-0e8a17318dbcd55054429182b74a5cc5R80
Can be part of `ForemanMaintain::Utils::CurlResponse` itself.

kgaikwad replied:
@swapab ,

Yes, I have noticed this before but I don't want to mix these changes in this PR.

Not only `port` but now we have introduced `rest-client` we don't even need `curl` here so that we can handle response in better way.
Once this PR get merged, I will modify this code as per new changes.

kgaikwad replied:
@swapab , @iNecas ,

Created new [issue](http://projects.theforeman.org/issues/20722) in Redmine to refactor existing DHCP config check as per new changes.

iNecas replied:
Why not to resolve the port part right away?

kgaikwad replied:
@iNecas ,
Yes, we can resolve port part now. But I thought of doing it in separate PR - 
convert curl-request `/dhcp` to `rest-client` request and automatically port part will get fixed.

swapab commented on definitions/features/foreman_proxy.rb:
```diff
@@ -1,21 +1,58 @@
+              :default_ssl_cert_options
 
   def initialize
     @dhcpd_conf_file = '/etc/dhcp/dhcpd.conf'
```
Should this be in config.yaml? OR Can we have a lookup helper for files? Instead of explicit declaration. Would be good to use across various distros.
Just a thought in case `.conf` files is located at some different locations.

swapab replied:
I feel there are lots of DHCP related calls viz. `valid_dhcp_configs?`, `dhcp_req_pass?`, `syntax_error_exists`.
A `DHCP` utility class can be initialized here.

kgaikwad replied:
@swapab ,
I don't think we require new `DHCP` utility class.
Instead I would prefer to modify `DhcpReservation` to `DHCP` as those APIs are nothing but categorized as `DHCP` APIs under `SmartProxy`.

There are different API categories under smart-proxy. 
Please check [URL](http://projects.theforeman.org/projects/smart-proxy/wiki/API) where smart-proxy APIs are listed, I hope this pretty much helpful.

Lets discuss about these topics in scrum meeting to make code better and clean.

swapab commented on definitions/procedures/dhcp/reservation/add.rb:
```diff
@@ -0,0 +1,20 @@
+      end
+
+      def run
+        feature(:foreman_proxy).add_reservation(@subnet, @ip, @mac, @name)
```
There should be a feature called `:dhcp_reservation` with `add/list/remove` methods on it.

swapab replied:
@iNecas Wanted to know is there a way to pass these arguments to the `initialize` method of the feature.
Another way to do would  be
`feature(:foreman_proxy, {:params => subnet: @subnet, ip: @ip, mac: @mac, name: @name}).add`

kgaikwad commented on lib/foreman_maintain/utils/foreman_proxy_settings.rb:
```diff
@@ -0,0 +1,32 @@
+        attrs.each do |key, value|
+          instance_variable_set("@#{key}", value)
+        end
+        #valid_attr_keys.each do |key|
```
I will delete this code.

swapab commented on config/foreman_maintain.yml.example:
```diff
@@ -17,5 +17,5 @@
-# Mention path where foreman-proxy certificates stored on filesystem
-# :foreman_proxy_cert_path: '/etc/foreman'
+# Mention path where foreman-proxy settings stored on filesystem
+# :foreman_proxy_settings_path,: '/etc/foreman-proxy/settings.yml'
```
Probably want to delete `,`

swapab commented on config/foreman_maintain.yml.packaging:
```diff
@@ -17,6 +17,5 @@
-# :foreman_proxy_cert_path: '/etc/foreman'
-
+# Mention path where foreman-proxy settings stored on filesystem
+# :foreman_proxy_settings_path,: '/etc/foreman-proxy/settings.yml'
```
Ditto.

kgaikwad commented:
@iNecas , @swapab 


For now, main DHCP command will also show same sub-commands as below. 
After changing name, it is creating an issue while running sub-command as available procedures are checked by using class hierarchy or label of procedure.

```
[root@satellite1 foreman_maintain]# ./bin/foreman-maintain dhcp --help
Usage:
    foreman-maintain dhcp [OPTIONS] SUBCOMMAND [ARG] ...

Parameters:
    SUBCOMMAND                    subcommand
    [ARG] ...                     subcommand arguments

Subcommands:
    dhcp-reservation-add          add DHCP reservation to subnet
    dhcp-reservation-delete       delete DHCP reservation using subnet
    dhcp-reservation-list         list DHCP reservations for subnet


Options:
    -h, --help                    print help
```

kgaikwad commented:
@iNecas , @swapab 

Fixed the test cases. Ready for review.

kgaikwad commented on definitions/features/foreman_proxy.rb:
```diff
@@ -1,72 +1,46 @@
-    success
+  def construct_foreman_proxy_url
+    foreman_url = settings.foreman_url
+    proxy_port = settings.https_port || settings.http_port
```
Yeah, make sense. I will add it to settings.

kgaikwad commented:
@swapab ,

Good Catch!!
I have added #execute method on dhcp command to print help text and modified code as per review comments.
Ready for review.

kgaikwad commented:
For check [verify_dhcp_config_syntax](foreman_maintain/definitions/checks/foreman_proxy_dhcp/verify_dhcp_config_syntax.rb),
`default` tag is replaced with new tag called `isp-dhcp-config-check` so that it won't be run in `health` as well as `upgrade` command.

upadhyeammit commented:
As per the foreman-maintain meeting we should have CRUD kind of operations should be more suitable into hammer tool as it already supports all of these kind of operations. As per this, closing this PR.
